### PR TITLE
Add functional test for java-spiffe-helper

### LINKED ISSUE #200: Add Functional Test for java-spiffe-helper
Add a functional test to the `java-spiffe-helper` module with the primary aim of detecting errors during CI. The test will verify the java-spiffe-helper's ability to connect to a SPIRE Agent and store certificates in a Java keystore.

Benefits:

- CI Error Detection: The primary benefit is the early detection of errors and regressions in the java-spiffe-helper during CI runs.
- Enhanced Reliability: Ensuring that the java-spiffe-helper consistently performs certificate management tasks.

### PR DESCRIPTION
**Pull request by moritzschmitz-oviva**
Closes #200 

1. Build the `java-spiffe-helper` container using the latest changes on the branch.
2. Bootstrap a `kind` cluster and load the previously built image into it.
3. Install the `spire-server` chart with dependencies. Set special values for the `CN` in the certificates. This is what we later test.
4. Start the `java-spiffe-helper` container and let it do its magic.
5. Copy the created `*.jks` files and verify the `CN` values set earlier.

Potential discussion topics:
1. Using the latest `spire` charts. Perhaps pin the version here?
2. Checking only for the `CN`. Anything else to check for? For example the properties are using hard-coded values. Could also make them dynamic.

### CODE REVIEW

File: .github/workflows/test.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+          tags: java-spiffe-helper:test
+          load: true
+          cache-from: type=gha
+          cache-to: type=gha,mode=max
+      - uses: helm/kind-action@v1
+        with:
+          cluster_name: kind
+      - run: kind load docker-image java-spiffe-helper:test --name kind
+      - run: helm upgrade --install -n spire-server spire-crds spire-crds --repo https://spiffe.github.io/helm-charts-hardened/ --create-namespace
+      - run: envsubst < .github/tests/spire-values.yaml.tpl > .github/tests/spire-values.yaml
```
Comment (kfox1111) [inline .github/workflows/test.yaml:31]: There's only two substitutions in the file, so could simplify it to just be static and override them in the helm install one line down adding:
  --set spire-server.ca_subject.common_name=$TRUSTSTORE_COMMON_NAME \
  --set spire-server.controllerManager.identities.clusterSPIFFEIDs.java-spiffe-helper.dnsNameTemplates[0]=$KEYSTORE_COMMON_NAME
  
  envsubst of a values file has always made me a little nervous.

**Review by kfox1111 — COMMENTED** [review]
This looks good to me. :)

One possible change inline.

**Comment by kfox1111** [comment]
Looks good to me. :)

File: .github/workflows/test.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+name: Test
```
Comment (maxlambrecht) [inline .github/workflows/test.yaml:1]:
```suggestion
name: Java SPIFFE Helper CI
```

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+          distribution: 'temurin'
+          java-version: '17'
+      - run: kubectl cp java-spiffe-helper:/tmp/keystore.p12 keystore.p12
+      - run: kubectl cp java-spiffe-helper:/tmp/truststore.p12 truststore.p12
+      - run: keytool -v -list -keystore keystore.p12 -storepass password | grep "CN=${{ env.KEYSTORE_COMMON_NAME }}"
+      - if: ${{ failure() }}
+        run: keytool -v -list -keystore keystore.p12 -storepass password
+      - run: keytool -v -list -keystore truststore.p12 -storepass password | grep "CN=${{ env.TRUSTSTORE_COMMON_NAME }}"
+      - if: ${{ failure() }}
+        run: keytool -v -list -keystore truststore.p12 -storepass password
```
Comment (maxlambrecht) [inline .github/workflows/test.yaml:59]: I'd suggest adding step names to improve clarity, something like:

```suggestion
- name: Capture Pod Logs on Failure
  if: ${{ failure() }}
  run: kubectl logs pod/java-spiffe-helper | tess java-spiffe-helper.log

- name: Upload Pod Logs on Failure
  if: ${{ failure() }}
  uses: actions/upload-artifact@v4
  with:
    name: java-spiffe-helper-logs
    path: java-spiffe-helper.log

- name: Describe Pod on Failure
  if: ${{ failure() }}
  run: kubectl describe pod/java-spiffe-helper

- name: Setup Java Environment
  uses: actions/setup-java@v4
  with:
    distribution: 'temurin'
    java-version: '17'

- name: Copy Keystore and Truststore from Pod
  run: |
    kubectl cp java-spiffe-helper:/tmp/keystore.p12 keystore.p12
    kubectl cp java-spiffe-helper:/tmp/truststore.p12 truststore.p12

- name: Verify Keystore Common Name
  run: keytool -v -list -keystore keystore.p12 -storepass ${{ secrets.KEYSTORE_PASSWORD }} | grep "CN=${{ env.KEYSTORE_COMMON_NAME }}"
  if: ${{ failure() }}
  run: keytool -v -list -keystore keystore.p12 -storepass ${{ secrets.KEYSTORE_PASSWORD }}

- name: Verify Truststore Common Name
  run: keytool -v -list -keystore truststore.p12 -storepass ${{ secrets.TRUSTSTORE_PASSWORD }} | grep "CN=${{ env.TRUSTSTORE_COMMON_NAME }}"
  if: ${{ failure() }}
  run: keytool -v -list -keystore truststore.p12 -storepass ${{ secrets.TRUSTSTORE_PASSWORD }}
```

File: .github/workflows/test.yaml [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+      - run: kind load docker-image java-spiffe-helper:test --name kind
+      - run: helm upgrade --install -n spire-server spire-crds spire-crds --repo https://spiffe.github.io/helm-charts-hardened/ --create-namespace
+      - run: |
+          helm upgrade --install -n spire-server spire spire \
+            --repo https://spiffe.github.io/helm-charts-hardened/ \
+            -f .github/tests/spire-values.yaml \
+            --set spire-server.ca_subject.common_name="$TRUSTSTORE_COMMON_NAME" \
+            --set spire-server.controllerManager.identities.clusterSPIFFEIDs.java-spiffe-helper.dnsNameTemplates[0]="$KEYSTORE_COMMON_NAME"
+      - run: kubectl apply -f .github/tests/java-spiffe-helper.yaml
+      - run: kubectl wait pod/java-spiffe-helper --for condition=Ready --timeout=90s
```
Comment (maxlambrecht) [inline .github/workflows/test.yaml:38]: This section could benefit from added clarity through the use of descriptive step names and breaking down complex commands for better readability:

Suggestion:

```suggestion
- name: Setup Kubernetes in Docker (KIND)
  uses: helm/kind-action@v1
  with:
    cluster_name: kind

- name: Load Docker Image into KIND
  run: kind load docker-image java-spiffe-helper:test --name kind

- name: Install SPIRE CRDs via Helm
  run: helm upgrade --install -n spire-server spire-crds spire-crds --repo https://spiffe.github.io/helm-charts-hardened/ --create-namespace

- name: Configure and Install SPIRE Server
  run: |
    helm upgrade --install -n spire-server spire spire \
      --repo https://spiffe.github.io/helm-charts-hardened/ \
      -f .github/tests/spire-values.yaml \
      --set spire-server.ca_subject.common_name="$TRUSTSTORE_COMMON_NAME" \
      --set spire-server.controllerManager.identities.clusterSPIFFEIDs.java-spiffe-helper.dnsNameTemplates[0]="$KEYSTORE_COMMON_NAME"

- name: Apply java-spiffe-helper Kubernetes Configuration
  run: kubectl apply -f .github/tests/java-spiffe-helper.yaml

- name: Wait for java-spiffe-helper Pod Readiness
  run: kubectl wait pod/java-spiffe-helper --for condition=Ready --timeout=90s
```

**Comment by maxlambrecht** [comment]
Hey @moritzschmitz-oviva, thanks a lot for this contribution! 

I've reviewed the changes and have a couple of suggestions:

- Could you please rename the `test.yaml` file to `java-spiffe-helper-ci.yaml`? This name better reflects its purpose within our CI pipeline.
- Also, it would be great if we could rename the folder from `github/workflows/tests` to `github/workflows/ci-k8s-configs`. 

In addition, I believe it's important to cover all supported LTS versions of Java to ensure compatibility. Could we update the job configuration to test against the following Java versions?

```suggestion
    name: Test on Java ${{ matrix.java_version }}
    runs-on: ubuntu-latest
    strategy:
      matrix:
        java_version: [8, 11, 17, 21]  # Including all current LTS versions 
```

**Comment by moritzschmitz-oviva** [comment]
> * Could you please rename the `test.yaml` file to `java-spiffe-helper-ci.yaml`? This name better reflects its purpose within our CI pipeline.

Done

> * Also, it would be great if we could rename the folder from `github/workflows/tests` to `github/workflows/ci-k8s-configs`.

Done

> In addition, I believe it's important to cover all supported LTS versions of Java to ensure compatibility. Could we update the job configuration to test against the following Java versions?

I don't think this is necessary. The `actions/setup-java` in the workflow is only required for `keytool`. `java-spiffe-helper` itself is containerized and built with the same `Dockerfile` which is used for the 'main' build as well.

**Comment by maxlambrecht** [comment]
> I don't think this is necessary. The `actions/setup-java` in the workflow is only required for `keytool`. `java-spiffe-helper` itself is containerized and built with the same `Dockerfile` which is used for the 'main' build as well.

You are completely right!

**Comment by maxlambrecht** [comment]
Hey @moritzschmitz-oviva, this PR looks good! I agree with the point you raised in the 'topics to discuss' about pinning the chart version. It's definitely a good practice that can enhance the stability of the test. Looking forward to seeing this adjustment.

**Comment by moritzschmitz-oviva** [comment]
> Hey @moritzschmitz-oviva, this PR looks good! I agree with the point you raised in the 'topics to discuss' about pinning the chart version. It's definitely a good practice that can enhance the stability of the test. Looking forward to seeing this adjustment.

Adjusted the workflow. Perhaps over-engineered it a bit now. Adding additional chart versions should be straight forward for testing. I didn't pin the patch, because according to SemVer, patches must be backwards compatible changes.

**Review by maxlambrecht — APPROVED** [review]
LGTM!

Thanks, @moritzschmitz-oviva !

**Pull request merged**